### ISSUE TITLE
How can I custom dataset

### ISSUE THREAD

**Issue by NhutTB** [cviviers/YOLOv5-6D-Pose#23]
Hi, thanks for your project it's helped me a lot.I have a question about customizing my dataset, your dataset LineMod uses ground truth with YOLO structure but dataset T-Less it's uses rotation and translation like this 
0:
- cam_R_m2c: [0.99813301, -0.01055917, 0.06015932, -0.00478128, -0.99542815, -0.09538872, 0.06089153, 0.09492298, -0.99362042]
  cam_t_m2c: [-3.08475792, 55.66261818, 595.96234477]
  obj_bb: [812, 803, 291, 292]
  obj_id: 1
1:
- cam_R_m2c: [0.99516573, 0.07745109, 0.06039457, 0.08295458, -0.99204393, -0.09469253, 0.05258004, 0.09924479, -0.99367318]
  cam_t_m2c: [-3.24573769, 55.52987781, 595.88827872]
  obj_bb: [813, 803, 291, 292]
  obj_id: 1

How can I turn it same with your data structure?
Thanks for reading.

**Comment by cviviers** [comment]
Hi @NhutTB 
These are the rotation and translation matrices. If you project the object onto the image using these you will get the keypoints to predict with this repo. 
You can then store the results as: 

- Centroid Coordinates: x0 (x-coordinate), y0 (y-coordinate)
- Corner Coordinates: From x1, y1 (first corner) to x8, y8 (eighth corner)
- Object Size Range: x range, y range
- Focal length x
- Focal length y
- Sensor width (in optical this can just be the image size)
- Sensor height (in optical this can just be the image size)
- Focal offset x (u0)
- Focal offset y (v0)
- Image width
- Image height
- Object rotation vector (Rodriques) [3x1] (this is the cam_R_m2c you have for T-Less, but in vector format) 
- Object translation vector [3x1] (this is the cam_t_m2c) you have above.

Have a look at the #[readme](https://github.com/cviviers/YOLOv5-6D-Pose/blob/main/data_curation/CreatingDataset.md) I made here for some more details.

**Comment by NhutTB** [comment]
Thanks for your reply.
I have a question about creating a dataset, I'm not qualified to use device that could return intrinsic matrix can I use Blender or similar tools to return an intrinsic matrix and rotation, translation matrices? As I know, we need Charuco board to detect translation and rotation, is this obligatory? And finally, I don't really understand this (Size encoding with 2 additional numbers representing the range in the x and y dimensions.),  Could you explain it in more detail and how I can get it?

PS: I suddenly have project 6d pose estimation so I don't have much knowledge about this, forgive me if I have an idiot question.

**Comment by NhutTB** [comment]
I have trouble with the data 
**Scanning '/content/drive/MyDrive/t-less_v2/train_canon/01/train' for images, labels, and masks... 14 found, 0 missing, 0 empty, 0 corrupted, 0 found, 14 missing, 0 empty, 0 corrupted:   1% 14/1036 [00:09<11:18,  1.51it/s]** 

**_https://drive.google.com/drive/folders/1gLj4lC_wRjkUEd4puZXSKRq3ncNdto9b?usp=sharing_**
This is my dataset, if you don't mind can you check it?

PS: I apologize for asking so many questions.

**Comment by cviviers** [comment]
Hey @NhutTB,

From a quick glance and the output you provide here, it seems the data is fine for keypoint prediction. I can't accuratly judge your camera intrinsic parameters though. It seems like a really large focal length. The camera's focal spot is also not close to the center of the image ( this is a bit high: 502.86220751). Especially if you made it in blender I expect the camera parameters to be perfect.

**Comment by NhutTB** [comment]
Hi, I have trouble training my custom dataset 

![Image](https://github.com/user-attachments/assets/6e8b5625-d017-4761-af05-50606049e521)

Why are there full 0%?

![Image](https://github.com/user-attachments/assets/16e3d160-1e88-4d23-a374-c0c9dc34483d)
-This is my image

1.000000 0.526400 0.537780 0.500000 0.500000 0.500000 0.383090 0.544721 0.500000 0.546682 0.383090 0.500000 0.739130 0.500000 0.592219 0.560763 0.739130 0.564440 0.592219 0.064440 0.356041 640.0 480.0 640 480 320.0 240.0 640 480
-This is my label

![Image](https://github.com/user-attachments/assets/42874db9-9cd7-4da3-9b59-d50e7b2935d0)
-This is my mask

![Image](https://github.com/user-attachments/assets/60dab8ea-227f-4111-b393-84cdca800217)
-This is my data structure

**Comment by NhutTB** [comment]
Hi @cviviers 

-Have my own dataset, but I want to use multiple same objects in 1 image, is it possible? If it's possible, how can I create .txt file in labels folder?

-Is it possible to use your model to detect it with my phone camera? In the state that the intrinsic matrix is not 100% true.

**Comment by cviviers** [comment]
In essence it is possible. It is exactly what the YOLO framework is set out to do. 
The current codebase is not perfectly set up to do that yet, but can easily be extended. 

Data preparation:
Simply just  add another row in the .txt file for each object in the image, even if they have the same class.
The camera intrinsic for each image will be the same, it is just the extrinsic for each object that is different.

Training loader:
The train loader should already be ready for this kind of data

Testing code:
This is the only part that might need some work. Currently the code picks the object with the highest confidence for each class. You will need to come up with some setup to pick the n-number of objects, depending on how many you expect. If you do not want to place a constraint on the number of objects of the same class, then you need to come up with a strategy of deciding when a prediction belongs to the same object, or is a different object - similar to non-maximum-suppression in object detection.

I hope this helps

**Comment by NhutTB** [comment]
Thanks for your advice, it's helping me a lot.

Now I have a problem when continue train from the last weight last.pt, it load more VRAM than training from scratch.
When I train with batch-size=16 from scratch, it takes 13.8 VRAM, but when I load the last.pt and continue training with batch-size=1, it takes 10.6 VRAM to train.

How can I fix this problem? 

I'm asking ChatGPT for help and this is its solution
Remove this line:

if ckpt['optimizer'] is not None:
    optimizer.load_state_dict(ckpt['optimizer'])

And this line too:

if ema and ckpt.get('ema'):
    ema.ema.load_state_dict(ckpt['ema'].float().state_dict())
    ema.updates = ckpt['updates']

Is it bad for model?

**Comment by NhutTB** [comment]
Hi @cviviers it's been a while. Sorry for late reply, I have other work to do.

I have 2 questions.
-First, how many epochs is enough? I try to train on 300 epochs and this is my test.

![Image](https://github.com/user-attachments/assets/474a27bf-d05a-422b-b643-13751df8c554)

It's great when I test on my blend data, but it's bad when I test object on real work.

![Image](https://github.com/user-attachments/assets/7765c6cc-349c-45fc-a1f7-2d563588e4bf)

-Second, can you take a quick look at my [dataset.](https://drive.google.com/drive/folders/132NYJLFWNG0yRNU-a_EmeUOF42hJ_HkB?usp=drive_link)

Is this enough for train data or do I need something else?

**Comment by cviviers** [comment]
Going from synthetic data to real can be quite tricky. Lighting conditions and texture on your object can play a big role, so you will need to apply the right kind of augmentation to reduce the domain gap. It will be great to add a few real images to the training set. You will also likely need them to validate how well your model is performing on real data. 

Check out the part of the tutorial on the Charuco board and how to create real labeled data.

Very cool to see the progress with your project!

**Comment by NhutTB** [comment]
Hi @cviviers, this is the first time I use Charuco board so can I ask you some questions?
- Do I need a specific size of Charuco board? Or just print test.pdf on an A4 paper.
- When placing the object on the Charuco board do I place it in the middle or anywhere as long as it is in the marker?

**Comment by cviviers** [comment]
- Yes, you need a matching software description of the board to what is printed. This means that if each marker is 5cm in software, it needs to be printed to 5 cm as well. GIMP has a printing function to set the size of the print.
- Anywhere on the board is fine, as long as the charuco board is successfully detected. typically 7 markers need to be accurately detected for the board to be detected, but more is better for robustness / accuracy of the pose
-  You need a checker board to calibrate the camera, and you need an aruco board for accurate poses - so charuco is best of both :)

**Comment by NhutTB** [comment]
It's been a while, and it's nice to see you. Today I have a question about data.
-In file yaml
![Image](https://github.com/user-attachments/assets/5cf0ca8d-7485-4376-94e8-1718b3e46f4f)

We need to input camera matrix, but in my data, I create data from 2 different sources. 1 from blender and 1 from charuco board.
But in 2 of the data I have a different camera matrix, how can I fix that?
Do I input 1 camera matrix or 2 camera matrices?

**Comment by cviviers** [comment]
Hi @NhutTB,

If you look at the updated linemod shared with the repo, you can see that the camera params are included in the label file for each image. This allows you to have different cameras per image. Just make sure to use those during training and inference.

**Comment by NhutTB** [comment]
Hi @cviviers, is it possible if i combine data from blender and real data, but real data I don't have mask images?
And have this error too.

![Image](https://github.com/user-attachments/assets/78225ee7-80e2-4f07-aeba-dc71f265e9cb)

**Comment by NhutTB** [comment]
Ahh, I think know this warn from, It's because I don't have enough mask images. I need to custom datasets.py or train.py to turn off that warning?

**Comment by NhutTB** [comment]
I can fix it now, just replace this in datasets.py

      if os.path.isfile(mk_file):
          nf_mask += 1  # mask found
          mk = Image.open(mk_file)
          mk.verify()
          if exif_size(mk) != shape:
              mk = np.zeros(shape, dtype=np.float32)
              assert exif_size(mk) == shape, 'mask does not fit image'
      else:
          nm_mask += 1  # mask missing
          mk = None
          mk_file = None

with this 

    if not os.path.isfile(mk_file):
        nm_mask += 1
        continue  
    
    try:
        nf_mask += 1  # mask found
        mk = Image.open(mk_file)
        mk.verify()
        if exif_size(mk) != shape:
            mk = np.zeros(shape, dtype=np.float32)
    except Exception as e:
        nc_mask += 1
        print(f'{prefix}WARNING: Ignoring corrupted mask {mk_file}: {e}')
        continue

**Comment by NhutTB** [comment]
Hi @cviviers, I have a question.

![Image](https://github.com/user-attachments/assets/22527f5c-4888-4cc2-8ed0-8089c17d11fc)

This line (train: Scanning '/content/YOLOv5-6D-Pose/Data-Weight/Synthetic_data/train.cache' for images, labels, masks... 1461 found, 0 missing, 0 empty, 0 corrupted, 1190 found, 271 missing, 0 empty, 0 corrupted: 100% 1461/1461 [00:00<?, ?it/s])

It's mean that I can train with 1461 images or just 1190 images that I have mask images ?

**Comment by vicky1626** [comment]
its possible to use real engine , real headlight for 6d pose  , how to create those dataset with  Charuco board ?
